### DISCUSSION TITLE
Provide user callback for custom implementation

### DISCUSSION THREAD

**Discussion by solution-loisir** [solution-loisir/markdown-it-eleventy-img#1] [Ideas]
Following a discussion with @AleksandrHovhannisyan, the idea of providing a user callback for custom implementation is really growing on me. It would go alongside of what's already develop. 

My main concern is about what arguments to include or should I include theme all? The markdown-it renderer signature include five parameters plus the plugin config objects (2) and the `Image` constructor which makes 8 arguments to pass to the callback, that's a lot.

Maybe we could drop the plugin config objects and be down to 6 arguments. This would mean passing the config object as `null` in `use`:

```js
config.setLibrary('md', markdownIt ()
  .use(markdownItEleventyImg, null, function {...}));
```
I need to think this over. I'm curious about others opinions...

**Comment by AleksandrHovhannisyan** [comment]
So the option I was thinking of is to have the plugin internally parse the image, read off all relevant attributes (`src`, `alt`, etc.), and then call `props.renderImage` with a plain JS object containing those attributes. So it would basically be a render function:

```js
.use(markdownItCustomImage, {
   // other plugin options can go here if you need them
  // ...
  renderImage: (imgData) => (),
}
```

I haven't looked at the code too carefully to see if this is doable, but it should be in theory. The only caveat is that it would require rebranding/republishing the plugin, or maybe publishing a second plugin for this use case. 

It might also be worth supporting custom `data-*` attributes (added via `markdown-it-attrs`), although I'm not sure how doable that would be.

This still wouldn't give me full feature parity with the shortcode approach, where I can pass in any prop I want on invocation to customize how my images behave (for example, if an image is above the fold, I might choose to disable the `lazy` attribute for _just that image_ so that Lighthouse doesn't flag me for lazily loading LCP images). That's currently doable with the shortcode approach but wouldn't be possible with this approach.

Just some stuff to consider!

**Comment by solution-loisir** [comment]
Hmm, I like the idea of passing a "tokenAttributes" object to the user callback. I think it could be a more advance use case for custom rendering and probably not incompatible with the actual plugin. Now, the callback would still need to have `Image` passed in so the user can implement their own rendering would it (nice to have)? As for the `loading="lazy"` attribute, I'm already using a `%lazy%` flag in the title string to control the presence of this attribute on the output. I can pass the `isLazy` (returned by the [findLazyFlag](https://github.com/solution-loisir/markdown-it-eleventy-img/blob/master/utilities/find-lazy-flag.js) function) property to the callback as well so it is totally doable. 

Here's what I understand from your input (feel free to correct):
```js
// Implementation of user callback option.

const Image = require("@11ty/eleventy-img");
const findLazyFlag = require("./utilities/find-lazy-flag");

module.exports = function markdownItEleventyImg(md, pluginOptions = {}, userCallback) {

  md.renderer.rules.image  = (tokens, index, rendererOptions, env, renderer) => {

    const token = tokens[index];

    const src = token.attrGet("src");
    const alt = token.content;
    const title = token.attrGet("title") || "";
    const { isLazy, titleText } = findLazyFlag(title);

    if(userCallback) {
      const tokenAttributes = {
        src,
        alt,
        title: titleText,
        isLazy
      }
      return userCallback(Image, tokenAttributes);
    }

    // Rest of the plugin code goes here...
  }
}
```
```js
// For users

.use(markdownItEleventyImg, null, (Image, tokenObject) => {...});
```
The aim of this project is to provide the ability to use modern images (which are tied to performance) in markdown while staying aligned with the [markdown philosophy](https://daringfireball.net/projects/markdown/syntax#philosophy). It is not intended to be a replacement for a shortcode. The idea is to keep the "easy-to-read and easy-to-write" feel of markdown. A use case that comes to mind would be images in an article written in markdown. It's probably nicer to write these images in markdown than with a shortcode. So the benefit is mainly ergonomic.

**Reply by AleksandrHovhannisyan** [reply]
Yup, that's basically what I had in mind! You could even make `userCallback` be part of `pluginOptions`.

I think making a plugin that doesn't have the 11ty image plugin as a dependency would be good, but this works 👍 

Also, just in case users want to support custom `data-*` attributes (like with the [markdown-it-attrs plugin](https://www.npmjs.com/package/markdown-it-attrs)), I think you could have the rendering function accept an optional third argument to fetch a specific attribute. Something like this should work:

```js
const getAttribute = (name) => token.attrGet(name);

// ...

return renderToImage(Image, imageAttributes, getAttribute);
```

Example usage:

```js
config.setLibrary("md", markdownIt().use(markdownItEleventyImg, {
    renderToImage: (ImagePlugin, imageAttributes, getAttribute) => {
        const customAttribute = getAttribute("data-whatever");
        const finalAttributes = { ...imageAttributes, customAttribute, etc };
        return ...
    }
}));
```

That way, most of the heavy lifting is already done for users, but in case they need to support a custom attribute, the plugin doesn't need to be updated to handle that use case.

**Reply by solution-loisir** [reply]
Well, those are great inputs! First of, I think the main point of this plugin is to process markdown images through eleventy-img as there seems to be a need for. Of course, an advance user deciding to produce her own render logic could decide to not use `Image` at all. Parameters are optional as usual.

Good idea to make `userCallback` be part of `pluginOptions`. I'll probably go that way.

I just refactored to use the `token.attrs` ([docs](https://markdown-it.github.io/markdown-it/#Token.prototype.attrs)) property instead of picking attributes one by one manually like I previously did using `token.attrGet(name)` (will be published in `v0.3.0`). With a little tweak, I can produce a dynamic object of all attributes passed to the image. Passing that object to the `userCallbak` will allow a user to have access to the full list of attributes included on the image. I tested with [markdown-it-attrs](https://www.npmjs.com/package/markdown-it-attrs) and it seems to be fully compatible (it's passing the custom attributes to the image). In my opinion, this is an upgrade from using `token.attrGet(name)` which can only get a specific attribute.

As for custom attributes, there is [`token.attrSet(name, value)`](https://markdown-it.github.io/markdown-it/#Token.attrSet) or [`token.attrPush([ name, value ])`](https://markdown-it.github.io/markdown-it/#Token.attrPush). those could be passed to `userCallback` like you describe, but I feel it would be more flexible to just pass `token` and let users implement what works best for their needs. Maybe something like this:
```js
const token = tokens[index];

// ...

return userCallback(Image, tokenAttributes, token);
```
This would have to be tested. Now, you don't necessarily need to modify the `token` to add custom attributes to the output. One important concept to understand is that this plugin is overriding the image rule renderer ([reference](https://markdown-it.github.io/markdown-it/#Renderer.prototype.rules)). So any string returned by the plugin will be the rendered markup of the image token (for every image tokens). In that context, adding custom attributes can be simple:
```js
config.setLibrary("md", markdownIt().use(markdownItEleventyImg, {
    render(Image, attrs, token) {
        return `<img src="${attrs.src}" alt="${attrs.alt}" data-framed>`
    }
}));
```
This would set a `data-framed` attribute on every markdown images. To be able to set custom attribute in markdown you have to do what markdown-it-attrs does: set custom delimiters in `md.core.ruler`, get the values and then set the attributes on the token before it renders. I would say this is outside of the scope of this plugin. Plus [markdown-it-attrs](https://www.npmjs.com/package/markdown-it-attrs) already does it and can be used side by side with this plugin (from `v0.3.0`) if needed. 

I need to think a bit more about this. Don't hesitate to share your ideas.

**Reply by AleksandrHovhannisyan** [reply]
> Passing that object to the userCallbak will allow a user to have access to the full list of attributes included on the image. I tested with [markdown-it-attrs](https://www.npmjs.com/package/markdown-it-attrs) and it seems to be fully compatible (it's passing the custom attributes to the image). In my opinion, this is an upgrade from using token.attrGet(name) which can only get a specific attribute.

Oh, awesome! That sounds like exactly what I want for my use case 👍

**Comment by solution-loisir** [comment]
@AleksandrHovhannisyan, just to let you know that I published `v0.4.0` with the first implementation of `renderImage`! Thanks again for all your inputs, very appreciated! If you want and have time to give it a spin (absolutely no pressure). I would like to ear from the user perspective. If you feel like it, Here's the [summary](https://www.npmjs.com/package/markdown-it-eleventy-img#custom-image-rendering) and [usage exemple](https://github.com/solution-loisir/markdown-it-eleventy-img/blob/master/.eleventy.js) (line 29). I'm off for a week with no internet so there might be delay in response. 😁

**Reply by AleksandrHovhannisyan** [reply]
Sounds good! I'll take a look when I get a chance.

**Comment by solution-loisir** [comment]
This feature has been implemented. Thanks for your implications!
